File: vllm_gcu/kernels/native_op/fused_qk_norm_rope.py
import torch
from vllm_gcu.kernels.native_op.utils import register_native
from vllm.model_executor.layers.layernorm import RMSNorm
from vllm.model_executor.layers.rotary_embedding import RotaryEmbedding

def _apply_qk_norm_rope(
    qkv: torch.Tensor,
    positions: torch.Tensor,
    q_norm: RMSNorm,
    k_norm: RMSNorm,
    rope: RotaryEmbedding,
    num_heads_q: int,
    num_heads_kv: int,
    head_dim: int,
) -> torch.Tensor:
    q_size = num_heads_q * head_dim
    kv_size = num_heads_kv * head_dim

    q, k, v = qkv.split([q_size, kv_size, kv_size], dim=-1)

    q_by_head = q.view(*q.shape[:-1], q.shape[-1] // head_dim, head_dim)
    q_by_head = q_norm.forward_native(q_by_head)
    q = q_by_head.view(q.shape)

    k_by_head = k.view(*k.shape[:-1], k.shape[-1] // head_dim, head_dim)
    k_by_head = k_norm.forward_native(k_by_head)
    k = k_by_head.view(k.shape)

    q, k = rope.forward_native(positions, q, k)
    return torch.cat([q, k, v], dim=-1)

@register_native("_C", "fused_qk_norm_rope")
def _ref_fused_qk_norm_rope(
    qkv,
    num_heads_q,
    num_heads_k,
    num_heads_v,
    head_dim,
    eps,
    q_weight,
    k_weight,
    cos_sin_cache,
    is_neox,
    position_ids,
):
    num_tokens = qkv.shape[0]
    num_heads = qkv.shape[1]

    positions = torch.arange(num_tokens, dtype=torch.long)
    q_norm = RMSNorm(head_dim, eps=eps, weight=q_weight)
    k_norm = RMSNorm(head_dim, eps=eps, weight=k_weight)

    rotary_dim = int(head_dim * 1.0)
    dtype = torch.bfloat16

    rope = RotaryEmbedding(
        head_size=head_dim,
        rotary_dim=rotary_dim,
        max_position_embeddings=4096,
        base=10000.0,
        is_neox_style=is_neox,
        dtype=dtype,
    )
    num_heads_kv = num_heads_k + num_heads_v
    head_dim = head_dim

    ref_result = _apply_qk_norm_rope(
        qkv,
        positions,
        q_norm,
        k_norm,
        rope,
        num_heads_q,
        num_heads_kv,
        head_dim,
    )

    qkv.copy_(ref_result)

File: csrc/src/fused_qk_norm_rope.cpp
/**
 * Copyright 2025 Enflame. All Rights Reserved.
 */
#include "fused_qk_norm_rope.h"

#include <topsaten/topsaten_vllm.h>

#include "tops_extension/torch/GCUAten.h"
#include "torch_gcu.h"

namespace vllm_gcu::llm_ops {
void fused_qk_norm_rope(at::Tensor &qkv, int64_t num_heads_q,
                        int64_t num_heads_k, int64_t num_heads_v,
                        int64_t head_dim, double eps, at::Tensor &q_weight,
                        at::Tensor &k_weight, at::Tensor &cos_sin_cache,
                        bool is_neox, at::Tensor &position_ids) {
  const torch_gcu::OptionalGCUGuard device_guard(device_of(qkv));
  const topsStream_t stream = torch_gcu::getCurrentGCUStream();
}
} // namespace vllm_gcu::llm_ops

